### PR TITLE
[main][Feature]Moe alltoallv communication optimization for unquantized RL training sence 

### PR DESCRIPTION
**Pull request by weijinqian0**
It comes from 0.9.1dev
[0.9.1][Feature]Moe alltoallv communication optimization for unquantized RL training sence & alltoallv support dpo (#1547)

- vLLM version: v0.10.0
- vLLM main: https://github.com/vllm-project/vllm/commit/97608dc276c292d9217eb6d334d969c5e89913c6

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -27,3 +27,5 @@ numba
 --pre
 --extra-index-url https://mirrors.huaweicloud.com/ascend/repos/pypi
 torch-npu==2.5.1.post1.dev20250619
+
+pytest_mock
```
Comment (wangxiyuan) [inline requirements.txt]: add this to requirement-dev.txt

File: tests/e2e/multicard/test_offline_inference_distributed.py [hunk 2]
```diff
@@ -137,6 +137,29 @@ def test_models_distributed_pangu():
...
+        "Briefly describe the major milestones in the development of artificial intelligence from 1950 to 2020.",
+        "Compare and contrast artificial intelligence with human intelligence in terms of processing information.",
+    ]
+    sampling_params = SamplingParams(max_tokens=5,
+                                     temperature=0.0,
+                                     top_k=50,
+                                     top_p=0.9)
+
+    with VllmRunner(
+            snapshot_download("vllm-ascend/pangu-pro-moe-pruing"),
```
Comment (wangxiyuan) [inline tests/e2e/multicard/test_offline_inference_distributed.py]: use pangu for test?

File: vllm_ascend/envs.py [hunk 3]
```diff
@@ -159,6 +159,11 @@
     # this feature is supported in A2, and eager mode will get better performance.
     "VLLM_ASCEND_ENABLE_MATMUL_ALLREDUCE":
     lambda: bool(int(os.getenv("VLLM_ASCEND_ENABLE_MATMUL_ALLREDUCE", '0'))),
+    # VLLM_ASCEND_ENABLE_MOE_ALL2ALL_SEQ:
```
Comment (wangxiyuan) [inline vllm_ascend/envs.py]: add more comment to describe what's this feature

File: vllm_ascend/models/qwen3_moe.py [hunk 4]
```diff
@@ -33,3 +36,4 @@ class CustomQwen3MoeForCausalLM(Qwen3MoeForCausalLM):
         "experts":
         ["experts.0.gate_proj", "experts.0.up_proj", "experts.0.down_proj"],
     }
+    qwen3.Qwen3MoeSparseMoeBlock = AscendSparseMoeBlock
```
Comment (wangxiyuan) [inline vllm_ascend/models/qwen3_moe.py]: this is not the correct way to rewreite model

File: vllm_ascend/ops/comm_utils.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+# This file is a part of the vllm-ascend project.
+import torch
+import torch.distributed
+import torch.distributed as dist
+import torch_npu
+
+COMM_STREAM = None
+
+
+def async_all_gather(input_,
```
Comment (wangxiyuan) [inline vllm_ascend/ops/comm_utils.py]: seems this func is not used by anywhere

File: vllm_ascend/ops/comm_utils.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+                                                        async_op=True)
+    else:
+        handle = torch.distributed._all_gather_base(ag_out,
+                                                    input_.contiguous(),
+                                                    group=group,
+                                                    async_op=True)
+    return input_, ag_out, handle
+
+
+def async_reduce_scatter(input_,
```
Comment (wangxiyuan) [inline vllm_ascend/ops/comm_utils.py]: ditto

File: vllm_ascend/ops/moe_dispatcher/token_dispatcher.py [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+                torch.npu.current_stream().synchronize()
+
+            return global_input_tokens
+
+        # token premute2 input
+        global_input_tokens = alltoall_token_permutation2(global_input_tokens)
+
+        return share_experts_output, global_input_tokens, tokens_per_expert
+
+    def preprocess_and_permtute1(self,
```
Comment (wangxiyuan) [inline vllm_ascend/ops/moe_dispatcher/token_dispatcher.py]: ditto

File: requirements.txt (hunk 1 above)
Comment (weijinqian0) [reply requirements.txt]: fixed

File: tests/e2e/multicard/test_offline_inference_distributed.py (hunk 2 above)
Comment (weijinqian0) [reply tests/e2e/multicard/test_offline_inference_distributed.py]: fixed

File: vllm_ascend/envs.py (hunk 3 above)
Comment (weijinqian0) [reply vllm_ascend/envs.py]: fixed

File: vllm_ascend/models/qwen3_moe.py (hunk 4 above)
Comment (weijinqian0) [reply vllm_ascend/models/qwen3_moe.py]: fixed

File: vllm_ascend/ops/comm_utils.py (hunk 5 above)
Comment (weijinqian0) [reply vllm_ascend/ops/comm_utils.py]: fixed

File: vllm_ascend/ops/comm_utils.py (hunk 6 above)
Comment (weijinqian0) [reply vllm_ascend/ops/comm_utils.py]: fixed

File: vllm_ascend/ops/moe_dispatcher/token_dispatcher.py (hunk 7 above)
Comment (weijinqian0) [reply vllm_ascend/ops/moe_dispatcher/token_dispatcher.py]: fixed

File: tests/ut/test_distributed_tensor_parallel.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
+#
```
Comment (MengqingCao) [inline tests/ut/test_distributed_tensor_parallel.py]: plz move this to `tests/distributed/test_tensor_paralle.py`

File: tests/ut/test_token_dispatcher.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
+#
```
Comment (MengqingCao) [inline tests/ut/test_token_dispatcher.py]: plz move this to `tests/moe_dispatcher/test_token_dispatcher.py`

File: vllm_ascend/ascend_forward_context.py [hunk 10]
```diff
@@ -18,6 +18,7 @@ class FusedMoEState(Enum):
     MC2 = 2
     AllGatherEP = 3
     NaiveMulticast = 4
+    All2AllSeq = 5
```
Comment (MengqingCao) [inline vllm_ascend/ascend_forward_context.py:21]: Maybe we could make more comments to describe clearly the differences between `All2AllSeq` and `All2All`

File: vllm_ascend/distributed/tensor_parallel.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# Adapts from: Megatron/megatron/core/tensor_parallel/mappings.py.
+# This file is a part of the vllm-ascend project.
+import torch
+
+
+def _gather_along_first_dim(input_, group, output_split_sizes=None):
```
Comment (MengqingCao) [inline vllm_ascend/distributed/tensor_parallel.py:21]: I think we can refactor the ops in this file into `distributed/communicator`, and call them through `get_tp_group().gather_along_first_dim` for example.

(Maybe the refactor could be done in the following up pr

**Review by MengqingCao — COMMENTED** [review]
Nit: we could do the following in the following up pr

File: vllm_ascend/distributed/tensor_parallel.py (hunk 11 above)
Comment (weijinqian0) [reply vllm_ascend/distributed/tensor_parallel.py:21]: ok, we'll consider it

File: vllm_ascend/ascend_forward_context.py (hunk 10 above)
Comment (weijinqian0) [reply vllm_ascend/ascend_forward_context.py:21]: There's some comments in envs.

File: tests/ut/test_distributed_tensor_parallel.py (hunk 8 above)
Comment (weijinqian0) [reply tests/ut/test_distributed_tensor_parallel.py]: ok

File: tests/ut/test_token_dispatcher.py (hunk 9 above)
Comment (weijinqian0) [reply tests/ut/test_token_dispatcher.py]: ok

File: tests/e2e/multicard/test_offline_inference_distributed.py [hunk 12]
```diff
@@ -157,6 +157,28 @@ def test_models_distributed_topk() -> None:
         vllm_model.generate(example_prompts, sampling_params)
 
 
+@patch.dict(os.environ, {"VLLM_ASCEND_ENABLE_MOE_ALL2ALL_SEQ": "1"})
+def test_models_distributed_alltoallv() -> None:
```
Comment (wangxiyuan) [inline tests/e2e/multicard/test_offline_inference_distributed.py:161]: This test is not ran  in CI. You should enable it here as well https://github.com/vllm-project/vllm-ascend/blob/main/.github/workflows/vllm_ascend_test.yaml#L277-L281

File: vllm_ascend/distributed/tensor_parallel.py (hunk 11 above)
Comment (wangxiyuan) [reply vllm_ascend/distributed/tensor_parallel.py:21]: I think this need GroupCoordinator to support gather_along_first_dim as well. It's need vLLM support it first. we can do it in the future.

**Review by wangxiyuan — COMMENTED** [review]
@jianzs  Please help review. Thanks. This is a large PR, while I think it's good to go, because it's a single feature and most of the code doesn't affect other function.

File: tests/ut/ops/test_token_dispatcher.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
+#
```
Comment (wangxiyuan) [inline tests/ut/ops/test_token_dispatcher.py:1]: this should be in ops/moe_dispatcher folder, it can be done in a follow up PR.

Comment (weijinqian0) [reply tests/ut/ops/test_token_dispatcher.py:1]: fixed

File: tests/e2e/multicard/test_offline_inference_distributed.py (hunk 12 above)
Comment (weijinqian0) [reply tests/e2e/multicard/test_offline_inference_distributed.py:161]: fixed

**Comment by wangxiyuan** [comment]
Let's merge this first to unblock other cherry-pick action. @jianzs any comment is welcome. It can be don in the follow up PR.

**Comment by jianzs** [comment]
> Let's merge this first to unblock other cherry-pick action. @jianzs any comment is welcome. It can be don in the follow up PR.

Got it.

**Comment by Yikun** [comment]
This PR break the Qwen3-30B-A3B accuracy test: https://github.com/vllm-project/vllm-ascend/actions/runs/16690820541/job/47248288106

```
 torch._dynamo.exc.Unsupported: dynamic shape operator: aten.bincount.default; Operator does not have a meta kernel that supports dynamic output shapes, please report an issue to PyTorch
 
 from user code:
    File "/__w/vllm-ascend/vllm-ascend/vllm-empty/vllm/model_executor/models/qwen3_moe.py", line 369, in forward
     hidden_states, residual = layer(positions, hidden_states, residual)
   File "/__w/vllm-ascend/vllm-ascend/vllm-empty/vllm/model_executor/models/qwen3_moe.py", line 313, in forward
     hidden_states = self.mlp(hidden_states)
   File "/__w/vllm-ascend/vllm-ascend/vllm_ascend/ops/fused_moe.py", line 1615, in forward
     hidden_states = self.experts(
   File "/__w/vllm-ascend/vllm-ascend/vllm_ascend/ops/fused_moe.py", line 1446, in forward
     e_hidden_states = self.quant_method.apply(
   File "/__w/vllm-ascend/vllm-ascend/vllm_ascend/ops/fused_moe.py", line 1159, in apply
     return fused_experts_with_all2all(hidden_states=x,
   File "/__w/vllm-ascend/vllm-ascend/vllm_ascend/ops/fused_moe.py", line 371, in fused_experts_with_all2all
     global_expert_tokens = torch.bincount(expanded_expert_idx,
 
 Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
 
 
 You can suppress this exception and fall back to eager by setting:
     import torch._dynamo
     torch._dynamo.config.suppress_errors = True
 
 
```

**Comment by Yikun** [comment]
> This PR break the Qwen3-30B-A3B accuracy test: https://github.com/vllm-project/vllm-ascend/actions/runs/16690820541/job/47248288106

issue: https://github.com/vllm-project/vllm-ascend/issues/2226
v0.9.1-dev: resolved since v0.9.1rc3: https://github.com/vllm-project/vllm-ascend/pull/2478
main: resolved since v0.10.0rc1: https://github.com/vllm-project/vllm-ascend/pull/2183
